### PR TITLE
Feature/heatmap

### PR DESCRIPTION
**Pull request by kvogli**
-Extended .proto file with the access points and heatmap gRPC and REST endpoints
-Implemented the interface functions in the server 
- [ ] Revert api dependency before merge

### CODE REVIEW

File: server/backend/rpcserver.go [hunk 1]
```diff
@@ -34,7 +34,7 @@ type CampusServer struct {
 }
 
 // Verify that CampusServer implements the pb.CampusServer interface
-var _ pb.CampusServer = (*CampusServer)(nil)
+// var _ pb.CampusServer = (*CampusServer)(nil)
```
Comment (joschahenningsen) [inline server/backend/rpcserver.go:37]: Whats the rationale behind this change?

File: server/backend/rpcserver.go [hunk 2]
```diff
@@ -47,6 +47,16 @@ func New(db *gorm.DB) *CampusServer {
...
+func (s *CampusServer) GetAccessPoint(ctx context.Context, in *pb.APRequest) (*pb.AccessPoint, error) {
+	log.Printf("Received request for AP with name: %s and timestamp: %s", in.Name, in.Timestamp)
+	return &pb.AccessPoint{Name: "response"}, nil
+}
+
+func (s *CampusServer) ListAccessPoints(in *pb.APRequest, stream pb.Campus_ListAccessPointsServer) error {
+	log.Println("Requesting list")
+	return nil
+}
+
```
Comment (joschahenningsen) [inline server/backend/rpcserver.go:59]: I assume these will be implemented later?

File: server/swagger/swagger.json [hunk 3]
```diff
@@ -1,5 +1,4 @@
 {
```
Comment (joschahenningsen) [inline server/swagger/swagger.json:1]: This file is not up to date, can you run `./generate.sh` in `/api` again? :)

File: api/go.mod [hunk 4]
```diff
@@ -11,10 +11,33 @@ require (
 )
 
 require (
+	github.com/bufbuild/buf v1.5.0 // indirect
```
Comment (joschahenningsen) [inline api/go.mod:14]: Can you run  `go mod tidy` in `/api`?

**Review by joschahenningsen — APPROVED** [review]

**Review by joschahenningsen — CHANGES_REQUESTED** [review]
Also there are a few merge conflicts, can you get the branch up to date?

**Comment by alexanderstephan** [comment]
@kvogli What is the current state on this?

**Comment by kvogli** [comment]
@alexanderstephan i rewrote the database queries and removed the injection threats as discussed. I also added 4 packages containing the server-side code of the heatmap and also a database under server/model/heatmap. I'm not sure where to place the python script (that is triggered daily) and the Dockerfile or the "extension" of the existing one - for this part of configuration i'm not really sure how to do.

**Comment by alexanderstephan** [comment]
@kvogli  

> rewrote the database queries and removed the injection threats as discussed. I also added 4 packages containing the server-side code of the heatmap and also a database under server/model/heatmap. 

Great!

> I'm not sure where to place the python script (that is triggered daily)

I guess putting it into the heatmap package itself is OK for now.


> Dockerfile or the "extension" of the existing one - for this part of configuration i'm not really sure how to do.

Not sure if I got this correctly. Just updating the existing Dockerfile in the root directory should suffice. It depends a bit on how many dependencies it introduces, but imo that should be fine.

**Comment by alexanderstephan** [comment]
Also please merge as soon possible, so it easier to review the changes.

**Comment by joschahenningsen** [comment]
@kvogli aside from merging with dev it might  be helpful to document the changes you made a little. e.g. 

- which endpoints and cronjobs does this PR implement? 
- How do they work? 
- Why is there a 7 MB sqlite file? Will it get get out of date? 
- Why is there a call to some python script (`forecasting.py`) that isn't in the repo? What should it do?
- Which migrations need to be done to the cronjob table? 
- Did you test your implementation against the context of the backend? Does it do what it's supposed to do?
- You `panic()` some times. This will crash the backend. Please return and handle the errors instead.
- Any reason you build custom sql queries instead of using gorm?
- go packages should be `lowercase`, please use this convention in `DBService`

Overall this PR needs more documentation on what it does, why it does it and how this is implemented.

File: server/model/heatmap/heatmap.go [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+type location struct {
+	lat  string
+	long string
+}
+
+var locations map[string]location
+
+func main() {
```
Comment (joschahenningsen) [inline server/model/heatmap/heatmap.go:155]: this is an entrypoint to an executable within the `models` package. Is this intended? This doesn't feel integrated in our current architecture.

**Comment by kvogli** [comment]
> @kvogli aside from merging with dev it might be helpful to document the changes you made a little. e.g.
> 
> * which endpoints and cronjobs does this PR implement?
> * How do they work?
> * Why is there a 7 MB sqlite file? Will it get get out of date?
> * Why is there a call to some python script (`forecasting.py`) that isn't in the repo? What should it do?
> * Which migrations need to be done to the cronjob table?
> * Did you test your implementation against the context of the backend? Does it do what it's supposed to do?
> * You `panic()` some times. This will crash the backend. Please return and handle the errors instead.
> * Any reason you build custom sql queries instead of using gorm?
> * go packages should be `lowercase`, please use this convention in `DBService`
> 
> Overall this PR needs more documentation on what it does, why it does it and how this is implemented.

* This PR imeplements 3 endpoints: "/accesspoints/heatmap/{name}", "/accesspoints/heatmap" and "/accesspoints/all-names".
* It would require 2 cronjobs, which should run daily for a few minutes each. One cronjob triggers the python script (which is now added in the repo) which does the forecasting on the number of connected devices of the access points, then updates the results in the database. The other cronjob also updates a table in the database. Another thing is that the python version needs to be 2.7 and not higher since a library depends on this version and the Prophet framework should also be installed. Both of these requirements are already achieved in this Docker image "alexcpn/fb_prophet_python:3.7.13-buster" so i'm not sure how to integrate it to the existing Docker container.
* The sqlite file is where all the data is stored needed for this feature. It has 3 tables.  It should not get out of date since it should be updated daily by the cronjobs.
* I'm not familiar with how the cronjob table is implemented - i would need help configuring it.
* I tested the implemtation agains the testserver and it worked as supposed to. I'm not sure if you ment smth else by context of backend?
* I started off with the usual sql queries and so ended up using them. I hope that is fine otherwise migrating them to gorm is a big rewrite.
* 'panic()' calls and package names are now handled.

This backend extension is supposed to serve another feature on the TUM Campus App for iOS. So it gets requests from the iOS client such as the history of the data of the access points or the predicted output. The webserver then queries the db and responds.

File: server/model/heatmap/heatmap.go (hunk 5 above)
Comment (kvogli) [reply server/model/heatmap/heatmap.go]: is now removed, was not intended

**Comment by joschahenningsen** [comment]
Thanks a lot for clarifying! I'll re- review the changes soon :)

File: server/backend/rpcserver.go (hunk 1 above)
Comment (kvogli) [reply server/backend/rpcserver.go]: is reverted.

File: server/backend/rpcserver.go (hunk 2 above)
Comment (kvogli) [reply server/backend/rpcserver.go]: done.

**Comment by COM8** [comment]
Closing this one for now since nobody has time right now to look into it further and fix stuff.

**Pull request closed without merging**